KAFKA-10223; Use NOT_LEADER_OR_FOLLOWER instead of non-retriable REPLICA_NOT_AVAILABLE for consumers

rajinisivaram opened the pull request:
Brokers currently return NOT_LEADER_FOR_PARTITION to producers and REPLICA_NOT_AVAILABLE to consumers if a replica is not available on the broker during reassignments. Non-Java clients treat REPLICA_NOT_AVAILABLE as a non-retriable exception, Java consumers handle this error by explicitly matching the error code even though it is not an InvalidMetadataException. This PR renames `NOT_LEADER_FOR_PARTITION` to `NOT_LEADER_OR_FOLLOWER` and uses the same error for producers and consumers. This is compatible with both Java and non-Java clients since all clients handle this error code (6) as retriable exception. The PR also makes ReplicaNotAvailableException a subclass of InvalidMetadataException.

- ALTER_REPLICA_LOG_DIRS continues to return REPLICA_NOT_AVAILABLE. Retained this for compatibility since this request never returned NOT_LEADER_FOR_PARTITION earlier.  Did not  deprecate ReplicaNotAvailableException because of this.
- MetadataRequest  version 0 also returns REPLICA_NOT_AVAILABLE for compatibility. Newer versions filter these out and
 return Errors.NONE, so didn't change this.
- Partition responses in MetadataRequest return REPLICA_NOT_AVAILABLE to indicate that one of the replicas is not available. Did not change this since NOT_LEADER_FOR_PARTITION is not suitable in this case.


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ijuma commented:
I think it makes sense to do this, but we also need to figure out a way to:
1. Make it clear that this is retriable for non Java clients via a KIP
2. Consider not returning this for older protocol versions.

hachikuji commented:
I'm not sure the REPLICA_NOT_AVAILABLE error is super helpful as distinguished from NOT_LEADER_FOR_PARTITION. One option here is just to use the latter in all cases.

ijuma commented:
I think the confusing part is returning NOT_LEADER_FOR_PARTITION if you are trying to fetch from a follower.

ijuma commented:
Maybe we could rename `NOT_LEADER_FOR_PARTITION`.

hachikuji commented:
I agree it's a little weird. Renaming NOT_LEADER is not a bad idea. I think what we want the error to convey is that the recipient is not a valid request target for the given partition. Maybe `INVALID_REQUEST_TARGET` or something like that.

rajinisivaram commented:
@ijuma @hachikuji  Thank you, I guess we need a KIP anyway even for renaming `NOT_LEADER_FOR_PARTITION`. We can discuss the options during KIP discussion. My only concern about renaming is that every one knows `NOT_LEADER_FOR_PARTITION` since it has been there forever, it may take a while to get used to a different name. Perhaps not an issue.

ijuma commented:
Renaming the error code name has no compatibility implications so in theory it could be done without a KIP. We could make this really boring and safe and call it `NOT_LEADER_OR_FOLLOWER`. I think this would not cause any confusion since the name is so similar and the length is similar since I trimmed the `FOR_PARTITION`.

ijuma commented:
If you both agree that the name above is reasonable, I would suggest sending a follow up to the KIP-320 thread covering what we've learned and how we're addressing it. It would be good to update the KIP text too so that we have something to refer to.

ijuma commented:
One issue is that the exception `NotLeaderForPartitionException` is in a public package. We could introduce a subclass, alwayys throw the subclass and deprecate `NotLeaderForPartitionException`.

rajinisivaram commented:
@ijuma Sounds good to me. Will see what @hachikuji thinks as well.

ijuma commented:
Also cc @edenhill as this affects non Java clients too.

ijuma commented:
We may also want to deprecate the error and exception for ReplicaNotAvailable since it would not serve any purpose (apart from clients retaining compatibility with some brokers), I think.

edenhill commented:
Renaming to NOT_LEADER_OR_FOLLOWER and reusing that error code in place of REPLICA_NOT_AVAILABLE sounds like a good idea. :+1:

hachikuji commented:
`NOT_LEADER_OR_FOLLOWER` sounds good to me. If we're looking to attach the change to a KIP, then KIP-392 might be a better option.

ijuma commented:
KIP-392 sounds good.

rajinisivaram commented:
@ijuma @hachikuji @edenhill Thanks for the responses. I will follow up on KIP-392 discussion thread and update the PR.

rajinisivaram committed:
KAFKA-10223; Replace REPLICA_NOT_AVAILABLE with NOT_LEADER_OR_FOLLOWER

rajinisivaram commented:
@ijuma @hachikuji Updated PR as discussed.

ijuma commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -121,8 +120,8 @@ class DelayedFetch(delayMs: Long,
+          case _: NotLeaderOrFollowerException =>  // Case A
+            debug(s"Broker is no longer the leader or follower of $topicPartition, satisfy $fetchMetadata immediately")
             return forceComplete()
           case _: ReplicaNotAvailableException =>  // Case B
```
Can we remove this now?

rajinisivaram replied:
yes, done.

ijuma commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -716,8 +712,12 @@ class ReplicaManager(val config: KafkaConfig,
+          case e: NotLeaderOrFollowerException =>
+            // Retaining REPLICA_NOT_AVAILABLE exception for ALTER_REPLICA_LOG_DIRS for compatibility
+            warn(s"Unable to alter log dirs for $topicPartition", e)
+            (topicPartition, Errors.REPLICA_NOT_AVAILABLE)
```
Would we want to change this for new IBPs?

rajinisivaram replied:
Updated to change to NOT_LEADER_OR_FOLLOWER from 2.7 onwards. Also added note to upgrade docs.

ijuma reviewed (commented):
Thanks for the PR. Some quick comments.

rajinisivaram committed:
Address review comments, remove some more usages of ReplicaNotAvailable

rajinisivaram commented:
@ijuma Thanks for the review. Another usage of REPLICA_NOT_AVAILABLE is in MetadataCache: https://github.com/rajinisivaram/kafka/blob/eaea63ec201fefd32256d75ed308aca311b42d1f/core/src/main/scala/kafka/server/MetadataCache.scala#L105 and a couple of lines below that one. These get returned for partition metadata in metadata responses and NOT_LEADER_OR_FOLLOWER doesn't look suitable. What do you think?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/errors/NotLeaderOrFollowerException.java:
```diff
@@ -0,0 +1,44 @@
+ * This could a transient exception during reassignments.
+ */
+@SuppressWarnings("deprecation")
+public class NotLeaderOrFollowerException extends NotLeaderForPartitionException {
```
One nitpick on the naming. It seems a little strange that followers might return `NOT_LEADER_OR_FOLLOWER` for requests which are intended for the leader. I think my suggestion to frame the name around the broker not being the intended target of the request was a little more accurate, though I admit the name was clumsier. Perhaps this class would be a good place to document the semantics? Maybe worth mentioning the Produce/Fetch behavior explicitly.

ijuma replied:
Good point.

rajinisivaram replied:
Updated javadoc.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -139,8 +139,8 @@
             LeaderNotAvailableException::new),
-    NOT_LEADER_FOR_PARTITION(6, "This server is not the leader for that topic-partition.",
-            NotLeaderForPartitionException::new),
+    NOT_LEADER_OR_FOLLOWER(6, "This server is not the leader or follower for that topic-partition.",
```
This would also be a good place to be clearer about the interpretation of this error. Maybe something like this?

> For requests intended only for the leader, this error indicates that the broker is not the current leader. For requests intended for any replica, this error indicates that the broker is not a replica of the topic partition.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -139,8 +139,8 @@
+            NotLeaderOrFollowerException::new),
     REQUEST_TIMED_OUT(7, "The request timed out.",
             TimeoutException::new),
     BROKER_NOT_AVAILABLE(8, "The broker is not available.",
```
Is it worth adding a note about the use of REPLICA_NOT_AVAILABLE below? I think we are effectively deprecating its usage in this PR.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -712,10 +708,13 @@ class ReplicaManager(val config: KafkaConfig,
+          case e: NotLeaderOrFollowerException =>
+            warn(s"Unable to alter log dirs for $topicPartition", e)
+            // Retaining REPLICA_NOT_AVAILABLE exception for ALTER_REPLICA_LOG_DIRS for older versions for compatibility
+            (topicPartition, if (config.interBrokerProtocolVersion >= KAFKA_2_7_IV0) Errors.NOT_LEADER_OR_FOLLOWER else Errors.REPLICA_NOT_AVAILABLE)
```
Hmm.. Wouldn't it make more sense to tie this to a version bump to this protocol rather than the IBP?

ijuma replied:
Yeah, this was a bad suggestion on my part as I didn't realize that we were using this from clients/tools versus for inter broker protocols. Maybe we should revert to using REPLICA_NOT_AVAILABLE here.

rajinisivaram replied:
Reverted.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@hachikuji @ijuma Thanks for the reviews, have addressed the comments so far.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java:
```diff
@@ -58,8 +58,8 @@
  * - {@link Errors#TOPIC_AUTHORIZATION_FAILED} If the user does not have READ access to a requested topic
- * - {@link Errors#REPLICA_NOT_AVAILABLE} If the request is received by a broker which is not a replica
- * - {@link Errors#NOT_LEADER_FOR_PARTITION} If the broker is not a leader and either the provided leader epoch
+ * - {@link Errors#REPLICA_NOT_AVAILABLE} If the request is received by a broker with version 2.4 to 2.6 which is not a replica
```
Is the note about the range from 2.4 to 2.6 correct? I think this error has always been possible in the case of a reassignment.

rajinisivaram replied:
Updated to say < 2.6.

hachikuji reviewed (approved):
LGTM. Just one minor comment.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -139,13 +139,15 @@
     BROKER_NOT_AVAILABLE(8, "The broker is not available.",
             BrokerNotAvailableException::new),
-    REPLICA_NOT_AVAILABLE(9, "The replica is not available for the requested topic-partition.",
+    REPLICA_NOT_AVAILABLE(9, "The replica is not available for the requested topic-partition. This is used for backward compatibility for " +
```
@hachikuji @rajinisivaram Since this is still used outside of produce/fetch, maybe the backwards compatibility message needs to be qualified?

rajinisivaram replied:
Updated the doc here and added javadoc for ReplicaNotAvailableException with more detail.

ijuma commented:
@rajinisivaram With regards to MetadataResponse, it's currently documented as:

```
/**
 * Possible topic-level error codes:
 *  UnknownTopic (3)
 *  LeaderNotAvailable (5)
 *  InvalidTopic (17)
 *  TopicAuthorizationFailed (29)

 * Possible partition-level error codes:
 *  LeaderNotAvailable (5)
 *  ReplicaNotAvailable (9)
 */
```

I don't think we should change it, but it does raise the question of whether we should make `ReplicaNotAvailable` retriable.

hachikuji commented:
I think it would make sense to change `ReplicaNotAvailableException` to extend `InvalidMetadataException`.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@hachikuji @ijuma Thanks for the reviews. I have updated the docs and made ReplicaNotAvailableException a subclass of InvalidMetadataException. Also moved upgrade doc under 2.6. If this doesn't go into 2.6, I will change all the javadocs and upgrade note to say 2.7.

ijuma commented on clients/src/main/java/org/apache/kafka/common/errors/ReplicaNotAvailableException.java:
```diff
@@ -16,7 +16,13 @@
+/**
+ * The replica is not available for the requested topic partition. This may be
+ * a transient exception during reassignments. From version 2.6 onwards, Fetch requests
+ * and other requests intended only for the leader of the topic partition return
```
Should it be `leader or follower` her and in the ReplicaNotAvailable message?

rajinisivaram replied:
@ijuma Thanks for the review, updated.

rajinisivaram committed:
Address review comment

ijuma reviewed (approved):
LGTM, thanks.

rajinisivaram commented:
@ijuma @hachikuji @bob-barrett Thanks for the reviews, merging to trunk and 2.6.

Nevon commented:
I was recently made aware of this change, and am looking to make the corresponding change in the NodeJS client tulios/kafkajs#1155.

One thing I don't really understand about this change is how REPLICA_NOT_AVAILABLE could change to being a retriable error. What does this mean for backwards compatibility with clusters pre-2.6? Was it always supposed to be retriable, and this just fixed an oversight, or does it mean that we will now end up retrying an operation that actually isn't retriable on clusters from before 2.6?

rajinisivaram commented:
@Nevon REPLICA_NOT_AVAILABLE was a transient error during reassignments with older clusters as well and should have been marked retriable. But it was not being handled as a retriable error by clients apart from Java consumers which did handle this specially as a retriable error. With the changes in this PR, we don't return REPLICA_NOT_AVAILABLE to clients except in the cases listed in the PR description. This ensures that both Java and existing non-Java consumers without any changes retry transient errors during reassignment (now NOT_LEADER_OR_FOLLOWER).

Nevon commented:
That clears it right up. Thanks for the info!
